### PR TITLE
Fix pyinotify directory-based log-rotate

### LINKED ISSUE #1379: logpath issue: move logic "collecting of logfiles" from client to server resp. filter ...
Just a reminder for me (will too long provide a PR for this, waiting for 716-cs ... and #1346).

Should fix imho great problem with "collecting of logfiles" client side: current solution uses fixedly list of the log files, found by the moment of the fail2ban starting. 
Variable logfiles of some services (multiple domains, etc.) will not be taken into account, as long as not restart of the fail2ban.
- [X] move logic "collecting of logfiles" from client to server resp. filter ... (* pro rata realized in my own branch);
- [ ] support appearing of new log files (matched) - (for example search in X minutes for new log files, or via pyinotify callback, if supported);
- [ ] better support disappearing of the log files (log-rotate, service down, etc...), and reactivating of jail from idle if log files found again;
- [ ] add `logregex` as additionally extended "replacement" for `logpath`;
- [ ] add `logregex-ignore` (and `logpath-ignore`) for negative filter;

### LINKED ISSUE #1769: ERROR Unable to get failures in
Hello. 
Periodically, an error occurs:

[~] tail -1 /logs/fail2ban.filter 
May  7 06:01:00 fail2ban.filter[14160]: ERROR Unable to get failures in /data/logs/linux/current

After the occurrence of this error, the processing of rules is terminated, until the fail2ban-client restarted. 

It seems that the filter tries to get the directory instead of the log file ('/data/logs/linux/current/sshd'), and why fail2ban does not try to open the file again and just stops the processing?

Thanks

***INFO***

[~] fail2ban-client -V
Fail2Ban v0.9.6

[~] cat /etc/redhat-release 
Red Hat Enterprise Linux Server release 6.8 (Santiago)

[~] uname -a
Linux 2.6.32-642.el6.x86_64 #1 SMP Wed 

[~] fail2ban-client -d
['set', 'syslogsocket', 'auto']
['set', 'loglevel', 'INFO']
['set', 'logtarget', 'SYSLOG']
['set', 'dbfile', '/var/lib/fail2ban/fail2ban.sqlite3']
['set', 'dbpurgeage', 86400]
['add', 'sshd', 'auto']
['set', 'sshd', 'usedns', 'warn']
['set', 'sshd', 'addlogpath', '/data/logs/linux/current/sshd', 'head']
['set', 'sshd', 'maxretry', 8]
['set', 'sshd', 'addignoreip', '127.0.0.1/8']
['set', 'sshd', 'logencoding', 'auto']
['set', 'sshd', 'bantime', 1801]
['set', 'sshd', 'ignorecommand', '']
['set', 'sshd', 'findtime', 600]
['set', 'sshd', 'maxlines', '10']

### PR DESCRIPTION
**Pull request by sebres**
* Fixed detection of directory-based log-rotation of pyinotify backend.
If directory moved and the target is not watched path, so the monitoring of it could not be continued.
Now fixed with pending files await a monitoring if there (resp. its directories) appears again (respawn).

* Differentiate between watched directories and files (refreshing monitoring of files/dirs expected different flags for watcher).

* Switch from ThreadedNotifier to Notifier:
  - instance of Filter is already a thread;
  - avoid stop of pyinotify processing if an interim error occurs (and breaks main-loop, e. g. during multi-threaded processing by add/remove log-files)

This logic (pending restoration) can be used in observer-thread by the implementation to fix #1379

Closes #1769

### CODE REVIEW

**Comment by sebres** [comment]
I've no interest to back-port this to 0.9 (because branches just too differently), if someone wants - welcome with PR.

File: fail2ban/server/filter.py [hunk 1]
```diff
@@ -895,7 +895,8 @@ def getFailures(self, filename):
 			# see http://python.org/dev/peps/pep-3151/
 			except IOError as e:
 				logSys.error("Unable to open %s", filename)
-				logSys.exception(e)
+				if e.errno != 2:
```
Comment (yarikoptic) [inline fail2ban/server/filter.py]: What is special about 2? Please leave a comment describing

File: fail2ban/server/filterpyinotify.py [hunk 2]
```diff
@@ -113,14 +151,81 @@ def _process_file(self, path):
 			self.failManager.cleanup(MyTime.time())
 		self.__modified = False
 
+	def _addPending(self, path, reason, isDir=False):
+		if path not in self.__pending:
+			self.__pending[path] = [self.sleeptime / 10, isDir];
```
Comment (yarikoptic) [inline fail2ban/server/filterpyinotify.py]: Why by 10? Now that in some obscure case could become 0, so much be worth ceiling float division instead

File: fail2ban/server/filterpyinotify.py [hunk 3]
```diff
@@ -113,14 +151,81 @@ def _process_file(self, path):
...
+			self.__pendingNextTime = 0
+			if isinstance(reason, pyinotify.Event):
+				reason = [reason.maskname, reason.pathname]
+			logSys.log(logging.MSG, "Log absence detected (possibly rotation) for %s, reason: %s of %s",
+				path, *reason)
```
Comment (yarikoptic) [inline fail2ban/server/filterpyinotify.py]: reason could be that event passed in... Is it a named tuple of guaranteed length 2?

File: fail2ban/server/filterpyinotify.py [hunk 4]
```diff
@@ -113,14 +151,81 @@ def _process_file(self, path):
...
+
+	def _delPending(self, path):
+		try:
+			del self.__pending[path]
+		except KeyError: pass
+
+	def _checkPending(self):
+		if self.__pending:
```
Comment (yarikoptic) [inline fail2ban/server/filterpyinotify.py]: A matter of style of cause but I would have just checked  for not and returned if so to not cause this massive indented block

File: fail2ban/server/filter.py (hunk 1 above)
Comment (sebres) [inline fail2ban/server/filter.py]: 2 - no file, but OK :)

File: fail2ban/server/filterpyinotify.py (hunk 2 above)
Comment (sebres) [inline fail2ban/server/filterpyinotify.py]: although sleeptime is float, but to be sure - yes, I'll do.

File: fail2ban/server/filterpyinotify.py (hunk 3 above)
Comment (sebres) [inline fail2ban/server/filterpyinotify.py]: either event (then transformed above) or array with 2 elms.

File: fail2ban/server/filterpyinotify.py (hunk 4 above)
Comment (sebres) [inline fail2ban/server/filterpyinotify.py]: No problem.

**Comment by sebres** [comment]
I found another more or less grave error there, that may sporadic occur within notifyier (by processing some events and modifying of watcher at the same time in another thread).
This should be fixed now within [62e580b7cf647a2f66c19a833e5aa703ad1c19f8].

**Pull request merged**